ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/ListIndexTool.java

```diff
@@ -0,0 +1,576 @@
+                // 1) the deleted index was explicitly passed as parameter to the /_cat/indices request. In this case we
+                // want the subsequent requests to fail.
+                // 2) the deleted index was resolved as part of a wildcard or _all. In this case, we want the subsequent
+                // requests not to fail on the deleted index (as we want to ignore wildcards that cannot be resolved).
+                // This behavior can be ensured by letting the cluster state, cluster health and indices stats requests
+                // re-resolve the index names with the same indices options that we used for the initial cluster state
+                // request (strictExpand).
+                sendClusterStateRequest(indices, subRequestIndicesOptions, local, client, new ActionListener<>() {
+                    @Override
+                    public void onResponse(ClusterStateResponse clusterStateResponse) {
+                        // Starts to fetch index details here, if a batch fails build whatever we have and return.
+                        fetchPages(
+                            indices,
+                            local,
+                            DEFAULT_CLUSTER_MANAGER_NODE_TIMEOUT,
+                            includeUnloadedSegments,
+                            pageParams,
+                            pageResults,
+                            clusterStateResponse,
+                            getSettingsResponse,
+                            subRequestIndicesOptions,
+                            originalListener
+                        );
+                    }
+
+                    @Override
+                    public void onFailure(final Exception e) {
+                        originalListener.onFailure(e);
+                    }
+                });
+            }
+
+            @Override
+            public void onFailure(final Exception e) {
+                originalListener.onFailure(e);
+            }
+        });
+    }
+
+    private void fetchPages(
+        String[] indices,
+        boolean local,
+        TimeValue clusterManagerNodeTimeout,
+        boolean includeUnloadedSegments,
+        PageParams pageParams,
+        Queue<Collection<ActionResponse>> pageResults,
+        ClusterStateResponse clusterStateResponse,
+        GetSettingsResponse getSettingsResponse,
+        IndicesOptions subRequestIndicesOptions,
+        ActionListener<Table> originalListener
+    ) {
+        final ActionListener<PageToken> iterativeListener = ActionListener.wrap(r -> {
+            // when previous response returns, build next request with response and invoke again.
+            PageParams nextPageParams = new PageParams(r.getNextToken(), pageParams.getSort(), pageParams.getSize());
+            // when next page doesn't exist or reaches max supported page size, return.
+            if (r.getNextToken() == null || pageResults.size() >= MAX_SUPPORTED_LIST_INDICES_PAGE_SIZE) {
```

Comment:
Right now we are pulling pages one by one. Can we parallelize this one? Something like this maybe?


```
private void fetchPagesInParallel(
    String[] indices,
    boolean local,
    boolean includeUnloadedSegments,
    PageParams initialPageParams,
    ClusterStateResponse clusterStateResponse,
    GetSettingsResponse getSettingsResponse,
    ActionListener<Table> listener
) {
    // Thread-safe collection for storing page results
    Queue<Collection<ActionResponse>> pageResults = new ConcurrentLinkedQueue<>();

    // ExecutorService to manage parallel threads
    ExecutorService executor = Executors.newFixedThreadPool(5); // Adjust concurrency as needed

    // Start fetching pages
    fetchPageRecursively(indices, local, includeUnloadedSegments, initialPageParams, clusterStateResponse, getSettingsResponse, pageResults, executor)
        .whenComplete((ignored, ex) -> {
            executor.shutdown(); // Shutdown the thread pool
            if (ex != null) {
                log.error("Error during parallel page fetching", ex);
                listener.onFailure(ex);
            } else {
                Table table = buildTable(clusterStateResponse, getSettingsResponse, pageResults);
                listener.onResponse(table);
            }
        });
}

private CompletableFuture<Void> fetchPageRecursively(
    String[] indices,
    boolean local,
    boolean includeUnloadedSegments,
    PageParams currentPageParams,
    ClusterStateResponse clusterStateResponse,
    GetSettingsResponse getSettingsResponse,
    Queue<Collection<ActionResponse>> pageResults,
    ExecutorService executor
) {
    return CompletableFuture.runAsync(() -> {
        fetchSinglePage(
            indices,
            local,
            includeUnloadedSegments,
            currentPageParams,
            clusterStateResponse,
            getSettingsResponse,
            pageResults
        );
    }, executor).thenCompose(ignored -> {
        // Check if there's a next page to fetch
        PageToken nextToken = getNextPageToken(currentPageParams, pageResults);
        if (nextToken != null) {
            PageParams nextPageParams = new PageParams(nextToken, currentPageParams.getSort(), currentPageParams.getSize());
            return fetchPageRecursively(indices, local, includeUnloadedSegments, nextPageParams, clusterStateResponse, getSettingsResponse, pageResults, executor);
        } else {
            return CompletableFuture.completedFuture(null); // No more pages to fetch
        }
    });
}

private void fetchSinglePage(
    String[] indices,
    boolean local,
    boolean includeUnloadedSegments,
    PageParams pageParams,
    ClusterStateResponse clusterStateResponse,
    GetSettingsResponse getSettingsResponse,
    Queue<Collection<ActionResponse>> pageResults
) {
    try {
        GroupedActionListener<ActionResponse> groupedListener = createGroupedListener(
            pageResults,
            pageParams.getRequestedToken(),
            ActionListener.wrap(
                response -> log.info("Successfully fetched page {}", pageParams.getRequestedToken()),
                error -> log.error("Failed to fetch page {}", pageParams.getRequestedToken(), error)
            )
        );

        // Fetch index stats
        sendIndicesStatsRequest(indices, IndicesOptions.lenientExpandHidden(), includeUnloadedSegments, ActionListener.wrap(
            groupedListener::onResponse,
            groupedListener::onFailure
        ));

        // Fetch cluster health
        sendClusterHealthRequest(indices, IndicesOptions.lenientExpandHidden(), local, DEFAULT_CLUSTER_MANAGER_NODE_TIMEOUT, client, ActionListener.wrap(
            groupedListener::onResponse,
            groupedListener::onFailure
        ));
    } catch (Exception e) {
        log.error("Unexpected error while fetching a single page", e);
    }
}

private PageToken getNextPageToken(PageParams pageParams, Queue<Collection<ActionResponse>> pageResults) {
    // Determine the next token based on current page's results or null if no more pages
    // Replace this with actual logic to extract the next token
    return pageResults.isEmpty() ? null : new PageToken(); // Placeholder logic
}
```

Reply:
No we can't as we need to get the page token from previous response and use it in next request.

Reply:
> Right now we are pulling pages one by one. Can we parallelize this one?

The whole purpose of using `_list` instead of `_cat` is to reduce CPU load on large clusters (hundreds of nodes) with a large number of indices (tens of thousands).   See 
 - https://github.com/opensearch-project/OpenSearch/issues/15954
 - https://github.com/opensearch-project/OpenSearch/pull/15986

Parallelizing would defeat the purpose of using `_list` -- just use `_cat` if you want the minimum latency but it comes at a cost of CPU.  The `_list` API intentionally trades away latency (for very large responses) in order to spread CPU out over a longer time period via pagination.